content_api/auth/auth_data_manager.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8; -*-
+#
+# This file is part of Superdesk.
+#
+# Copyright 2013, 2014, 2015 Sourcefabric z.u. and contributors.
+#
+# For the full copyright and license information, please see the
+# AUTHORS and LICENSE files distributed with this source code, or
+# at https://www.sourcefabric.org/superdesk/license
+
+from datetime import datetime, timedelta
+
+from flask import current_app as app
+from redis.client import StrictRedis
+
+import superdesk
+
+
+redis = StrictRedis()
+
+
+class DictObject:
+    def __init__(self, **args):
+        self.__dict__.update(args)
+
+
+class AuthDataManager:
+
+    @staticmethod
+    def get_client(client_id):
+        """ Loads a client from mongodb and returns it as a Client or None.
+        """
+        client = superdesk.get_resource_service('clients').find_one(req=None, _id=client_id)
+        client['client_id'] = client['_id']
+        client['default_scopes'] = app.config['OAUTH2_SCOPES']
+        client['default_redirect_uri'] = app.config['PUBLICAPI_URL']
+        if not client:
+            return None
+        return DictObject(**client)
+
+    @staticmethod
+    def get_user(username, password, *args, **kwargs):
+        """ Loads a user from mongodb and returns it as a User or None.
+        """
+        user = superdesk.get_resource_service('users').find_one(req=None, username=username)
+        if not user:
+            return None
+        elif password and user['password'] != password:
+            return None
+        return DictObject(**user)
+
+    @staticmethod
+    def get_token(access_token=None, refresh_token=None):
+        """ Loads a token from mongob and returns it as a Token or None.
+        """
+        if not (access_token or refresh_token):
+            return None
+
+        if access_token:
+            token = superdesk.get_resource_service('tokens').find_one(req=None, access_token=access_token)
+        elif refresh_token:
+            token = superdesk.get_resource_service('tokens').find_one(req=None, refresh_token=refresh_token)
+        return DictObject(token)
```

Comment:
Sure, I'll just create a new ticket so that we don't forget about it.

BTW, just for the reference - the network latency and other random factors indeed make it extremely difficult to detect a tiny difference in request processing times. On the other hand, though, sending enough requests (might be as low as just _hundreds_ of them) makes the random noise to statistically evens out enough to make the difference in average processing times noticeable and accurate.

This link explains it very well - http://codahale.com/a-lesson-in-timing-attacks/